File: claude_exp/routers/downstream_host/downstream_shared/stats.py
import threading
import time
from collections import deque
from datetime import datetime

_WINDOWS = (30, 60, 180, 1800)

# Bounded by count, not time - unlike _sent_times/_recv_times (which need real timestamps to
# answer "how many in the last 30s"), a percentile only needs *some* recent sample of values, and
# a fixed-size deque keeps memory and the snapshot()-time sort cost bounded even at high TPS
# (sorting a couple thousand floats is sub-millisecond; sorting an unbounded hours-long window
# would not be). Same "keep last N" convention as shared/log_buffer.py's LogBuffer.
_LATENCY_MAXLEN = 2000


def _percentile(sorted_values: list, pct: float) -> float:
    """Linear-interpolation percentile - no numpy dependency for what's otherwise a very small,
    already-dependency-light shared module."""
    if not sorted_values:
        return None
    k = (len(sorted_values) - 1) * pct
    f = int(k)
    c = min(f + 1, len(sorted_values) - 1)
    if f == c:
        return sorted_values[f]
    return sorted_values[f] + (sorted_values[c] - sorted_values[f]) * (k - f)


class Stats:
    def __init__(self, yellow_threshold_seconds=None):
        self._lock = threading.Lock()
        self._yellow_threshold_seconds = yellow_threshold_seconds
        self._sent_total = 0
        self._recv_total = 0
        self._sent_times = deque()
        self._recv_times = deque()
        self._last_recv_time = None
        self._connections = {}
        self._gauges = {}
        self._latencies = {}

    def set_connection(self, name: str, connected: bool) -> None:
        with self._lock:
            self._connections[name] = bool(connected)

    def set_gauge(self, name: str, value) -> None:
        with self._lock:
            self._gauges[name] = value

    def record_sent(self) -> None:
        with self._lock:
            self._sent_total += 1
            self._sent_times.append(time.time())
            self._trim(self._sent_times)

    def record_recv(self) -> None:
        with self._lock:
            now = time.time()
            self._recv_total += 1
            self._recv_times.append(now)
            self._trim(self._recv_times)
            self._last_recv_time = now

    def record_latency(self, name: str, value_ms: float) -> None:
        """Records one timing sample under a named bucket (e.g. "queue_wait", "crypto_rtt",
        "downstream_rtt", "total" - see router/dispatcher.py) for live per-hop latency
        visibility in /stats, without waiting on an offline soak-test CSV. Available on every
        actor type (Stats is shared), not just the router - any actor with a hop worth timing
        can call this the same way."""
        with self._lock:
            bucket = self._latencies.get(name)
            if bucket is None:
                bucket = deque(maxlen=_LATENCY_MAXLEN)
                self._latencies[name] = bucket
            bucket.append(value_ms)

    @staticmethod
    def _trim(times: deque) -> None:
        cutoff = time.time() - max(_WINDOWS)
        while times and times[0] < cutoff:
            times.popleft()

    @staticmethod
    def _count_within(times: deque, window_seconds: int) -> int:
        cutoff = time.time() - window_seconds
        count = 0
        for t in reversed(times):
            if t < cutoff:
                break
            count += 1
        return count

    def snapshot(self) -> dict:
        with self._lock:
            result = {"sent_total": self._sent_total, "recv_total": self._recv_total}
            for window in _WINDOWS:
                result[f"sent_{window}s"] = self._count_within(self._sent_times, window)
                result[f"recv_{window}s"] = self._count_within(self._recv_times, window)

            if self._last_recv_time is not None:
                result["seconds_since_last_recv"] = round(time.time() - self._last_recv_time, 1)
                result["last_recv_datetime"] = datetime.fromtimestamp(self._last_recv_time).strftime("%H:%M:%S")
            else:
                result["seconds_since_last_recv"] = None
                result["last_recv_datetime"] = None

            if self._yellow_threshold_seconds is not None:
                result["yellow_threshold_seconds"] = self._yellow_threshold_seconds

            if self._connections:
                result["connections"] = dict(self._connections)

            if self._gauges:
                result["gauges"] = dict(self._gauges)

            if self._latencies:
                latency_out = {}
                for name, bucket in self._latencies.items():
                    if not bucket:
                        continue
                    values = sorted(bucket)
                    latency_out[name] = {
                        "count": len(values),
                        "min_ms": round(values[0], 3),
                        "p50_ms": round(_percentile(values, 0.50), 3),
                        "p95_ms": round(_percentile(values, 0.95), 3),
                        "max_ms": round(values[-1], 3),
                    }
                if latency_out:
                    result["latency"] = latency_out

            return result

File: claude_exp/routers/router_cpp/src/shared/stats.cpp
#include "shared/stats.h"

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <ctime>
#include <vector>

namespace shared {

namespace {

constexpr std::array<int, 4> kWindows = {30, 60, 180, 1800};
constexpr int kMaxWindowSeconds = 1800;

int64_t now_ms() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
               std::chrono::system_clock::now().time_since_epoch())
        .count();
}

double round3(double v) { return std::round(v * 1000.0) / 1000.0; }

// Linear-interpolation percentile - matches router_py's _percentile helper.
double percentile(const std::vector<double>& sorted_values, double pct) {
    double k = (sorted_values.size() - 1) * pct;
    size_t f = static_cast<size_t>(k);
    size_t c = std::min(f + 1, sorted_values.size() - 1);
    if (f == c) return sorted_values[f];
    return sorted_values[f] + (sorted_values[c] - sorted_values[f]) * (k - f);
}

}  // namespace

Stats::Stats(std::optional<int> yellow_threshold_seconds)
    : yellow_threshold_seconds_(yellow_threshold_seconds) {}

void Stats::set_connection(const std::string& name, bool connected) {
    std::lock_guard<std::mutex> lock(mutex_);
    connections_[name] = connected;
}

void Stats::set_gauge(const std::string& name, nlohmann::json value) {
    std::lock_guard<std::mutex> lock(mutex_);
    gauges_[name] = std::move(value);
}

void Stats::record_sent() {
    std::lock_guard<std::mutex> lock(mutex_);
    int64_t t = now_ms();
    sent_ts_.push_back(t);
    ++sent_total_;
    trim(sent_ts_, t);
}

void Stats::record_recv() {
    std::lock_guard<std::mutex> lock(mutex_);
    int64_t t = now_ms();
    recv_ts_.push_back(t);
    ++recv_total_;
    last_recv_ms_ = t;
    trim(recv_ts_, t);
}

void Stats::record_latency(const std::string& name, double value_ms) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto& bucket = latencies_[name];
    bucket.push_back(value_ms);
    while (bucket.size() > kLatencyMaxLen) {
        bucket.pop_front();
    }
}

void Stats::trim(std::deque<int64_t>& dq, int64_t now_ms_val) const {
    int64_t cutoff = now_ms_val - static_cast<int64_t>(kMaxWindowSeconds) * 1000;
    while (!dq.empty() && dq.front() < cutoff) {
        dq.pop_front();
    }
}

int Stats::count_within(const std::deque<int64_t>& dq, int64_t now_ms_val, int window_seconds) const {
    int64_t cutoff = now_ms_val - static_cast<int64_t>(window_seconds) * 1000;
    int count = 0;
    for (auto it = dq.rbegin(); it != dq.rend(); ++it) {
        if (*it >= cutoff) {
            ++count;
        } else {
            break;
        }
    }
    return count;
}

nlohmann::json Stats::snapshot() const {
    std::lock_guard<std::mutex> lock(mutex_);
    int64_t t = now_ms();

    nlohmann::json out;
    out["sent_total"] = sent_total_;
    out["recv_total"] = recv_total_;
    for (int w : kWindows) {
        out["sent_" + std::to_string(w) + "s"] = count_within(sent_ts_, t, w);
        out["recv_" + std::to_string(w) + "s"] = count_within(recv_ts_, t, w);
    }

    if (last_recv_ms_.has_value()) {
        double secs = static_cast<double>(t - *last_recv_ms_) / 1000.0;
        secs = std::round(secs * 10.0) / 10.0;
        out["seconds_since_last_recv"] = secs;

        std::time_t tt = static_cast<std::time_t>(*last_recv_ms_ / 1000);
        std::tm tm_buf{};
        localtime_r(&tt, &tm_buf);
        char buf[16];
        std::strftime(buf, sizeof(buf), "%H:%M:%S", &tm_buf);
        out["last_recv_datetime"] = std::string(buf);
    } else {
        out["seconds_since_last_recv"] = nullptr;
        out["last_recv_datetime"] = nullptr;
    }

    if (yellow_threshold_seconds_.has_value()) {
        out["yellow_threshold_seconds"] = *yellow_threshold_seconds_;
    }
    if (!connections_.empty()) {
        nlohmann::json conns = nlohmann::json::object();
        for (const auto& [k, v] : connections_) {
            conns[k] = v;
        }
        out["connections"] = std::move(conns);
    }
    if (!gauges_.empty()) {
        nlohmann::json g = nlohmann::json::object();
        for (const auto& [k, v] : gauges_) {
            g[k] = v;
        }
        out["gauges"] = std::move(g);
    }
    if (!latencies_.empty()) {
        nlohmann::json latency_out = nlohmann::json::object();
        for (const auto& [name, bucket] : latencies_) {
            if (bucket.empty()) continue;
            std::vector<double> values(bucket.begin(), bucket.end());
            std::sort(values.begin(), values.end());
            nlohmann::json row;
            row["count"] = values.size();
            row["min_ms"] = round3(values.front());
            row["p50_ms"] = round3(percentile(values, 0.50));
            row["p95_ms"] = round3(percentile(values, 0.95));
            row["max_ms"] = round3(values.back());
            latency_out[name] = std::move(row);
        }
        if (!latency_out.empty()) {
            out["latency"] = std::move(latency_out);
        }
    }
    return out;
}

}  // namespace shared
